### ISSUE TITLE
storage: node does not become live after restart until many hours later

### ISSUE THREAD

**Issue by jethrogb** [cockroachdb/cockroach#28179]
**BUG REPORT**

Please describe the issue you observed, and any steps we can take to reproduce it:

- Which version of CockroachDB are you using?
2.0.4

- What did you do?
Rebooted a node

- What did you expect to see?
Node to come up

- What did you see instead?
cockroachdb starts but node never becomes live

- What was the impact?
Minor, rest of the cluster is live

I'm seeing lots of messages like this:
```
W180801 21:55:27.492922 343 storage/node_liveness.go:501  [n17,hb] slow heartbeat took 4.5s
W180801 21:55:27.492930 343 storage/node_liveness.go:438  [n17,hb] failed node liveness heartbeat: context deadline exceeded
W180801 21:55:28.035125 344 sql/jobs/registry.go:300  canceling all jobs due to liveness failure
```

NTP sync and network connectivity seems good.

Log file for this node:
[cockroach.log](https://github.com/cockroachdb/cockroach/files/2251285/cockroach.log)

**Comment by jethrogb** [comment]
The log file contains an ASCII NUL character, could that be indicative of an issue?

**Comment by knz** [comment]
@tim-o can you qualify?

**Comment by tim-o** [comment]
@jethrogb - Do you see any metrics for this node appearing in the UI? From what I can see in the logs, it has actually joined the cluster, but the log rows indicate that it can't find itself on the liveness table. Which could mean that there's a network partition preventing the node from contacting the leaseholder of the liveness range. You can find out for sure by visiting:

<adminURL>/reports/range/n

Start with n=1 and walk forward until "Key Range" is "/System/NodeLiveness to /System/NodeLivenessMax". For a relatively fresh cluster this shouldn't take long (the liveness range on my local cluster is id 3).

Once you find it, note the lease holder and ensure that n17 is able to contact that node.

**Comment by jethrogb** [comment]
The node was finally able to join the cluster on its own. Log file: [cockroach.log](https://github.com/cockroachdb/cockroach/files/2258256/cockroach.log). Time of interest is `180802 03:25:03.838186`.

**Comment by tschottdorf** [comment]
@tim-o  I took a look at the logs and it seems that the proximate cause of this stall is something about the replica of the first range on this node (`r1/36`).

Going from top to bottom, we see this message:

```
W180801 21:48:43.945153 347 kv/dist_sender.go:1305  [range-lookup=/System/"system-version/",range-lookup=/Meta2/System/"system-version/\x00",n17] have been waiting 1m0s sending RPC to r1 (currently pending: [(n17,s17):36]) for batch: Scan [/Meta1/System/"system-version/\x00\x00",/Meta2/"")
```

This is trying to read a key used by the SQL lease manager, and basically takes a long time, which prevents the SQL subsystem from booting up. Why is this request blocked? Well, it's trying to send to r1. We have an optimization in DistSender that [prefers sending to the local node](https://github.com/tschottdorf/cockroach/blob/835cc098da0698ea9a12271d4e2998fad69a8e4a/pkg/kv/replica_slice.go#L146) (if it doesn't know where the leaseholder is, as would be the case for a freshly started node). Thus, we see attempts to acquire the lease on `r1/36`, but these also get stuck:

```
$ grep -aF 'r1/36' cr.log
W180801 21:48:43.903995 474 storage/replica.go:1502  [n17,s17,r1/36:/{Min-System/}] have been waiting 1m0s attempting to acquire lease
W180801 21:48:43.945067 349 storage/replica.go:1502  [n17,s17,r1/36:/{Min-System/}] have been waiting 1m0s attempting to acquire lease
W180801 21:48:43.946464 383 storage/replica.go:1502  [n17,s17,r1/36:/{Min-System/}] have been waiting 1m0s attempting to acquire lease
W180801 21:48:43.957101 61 storage/replica.go:1502  [n17,s17,r1/36:/{Min-System/}] have been waiting 1m0s attempting to acquire lease
W180801 21:49:13.946048 1037 storage/replica.go:1502  [n17,s17,r1/36:/{Min-System/}] have been waiting 1m0s attempting to acquire lease
W180802 03:25:03.676354 474 storage/store.go:1456  [n17,s17,r1/36:/{Min-System/}] could not gossip first range descriptor: [NotLeaseHolderError] r1: replica (n17,s17):36 not lease holder; lease holder unknown
W180802 11:30:51.084378 474 storage/store.go:1456  [n17,s17,r1/36:/{Min-System/}] could not gossip first range descriptor: [NotLeaseHolderError] r1: replica (n17,s17):36 not lease holder; lease holder unknown
I180802 20:16:44.828813 416 storage/replica.go:3290  [n17,s17,r1/36:/{Min-System/}] not quiescing: 1 pending commands
```

The liveness heartbeats would log a similar message (everything needs to access r1 once when the cluster is fresh, to populate the DistSender caches) if they got stuck for at least a minute, but they have a context timeout that ends them early after ~4.5s, resulting in the periodic log messages.

So why is r1 stuck? The easy explanation would be a loss of quorum, but I think it's implicit in the original post that the remaining nodes were active. The Gossip network shows 15 nodes as active, which I assume is the number of nodes in your cluster, @jethrogb? Based on that, looks like the range *should* have just worked.

```
W180802 03:25:03.676354 474 storage/store.go:1456  [n17,s17,r1/36:/{Min-System/}] could not gossip first range descriptor: [NotLeaseHolderError] r1: replica (n17,s17):36 not lease holder; lease holder unknown
```

^- is logged right before things unwedge, but I think that's a red herring. This goroutine was also blocked on a lease acquisition, but at least it tells us which error the lease acquisition unblocks with -- it's an odd one, because usually you get a lease or you find out who has it. Here we did neither. I think this might be the case of the lease getting rejected with a lease that isn't valid any more (in which case we'll [strip it](https://github.com/cockroachdb/cockroach/blob/ab1cb1f469dfe9692ef42689441791e574e5b40e/pkg/storage/replica.go#L1511)).

@jethrogb it would be helpful if you also provided the logs for the other nodes for this period of time, thanks!

**Comment by jethrogb** [comment]
Yes, 15 nodes in the cluster. The relevant logs for the other nodes:

[cockroach.srv4-ch.cockroachdb.2018-08-01T19_09_27Z.008509.log.gz](https://github.com/cockroachdb/cockroach/files/2267279/cockroach.srv4-ch.cockroachdb.2018-08-01T19_09_27Z.008509.log.gz)
[cockroach.srv4-dc.cockroachdb.2018-08-01T19_15_07Z.006175.log.gz](https://github.com/cockroachdb/cockroach/files/2267280/cockroach.srv4-dc.cockroachdb.2018-08-01T19_15_07Z.006175.log.gz)
[cockroach.srv4-sv.cockroachdb.2018-08-01T19_09_31Z.007232.log.gz](https://github.com/cockroachdb/cockroach/files/2267281/cockroach.srv4-sv.cockroachdb.2018-08-01T19_09_31Z.007232.log.gz)
[cockroach.srv5-ch.cockroachdb.2018-08-01T20_25_35Z.012026.log.gz](https://github.com/cockroachdb/cockroach/files/2267282/cockroach.srv5-ch.cockroachdb.2018-08-01T20_25_35Z.012026.log.gz)
[cockroach.srv5-dc.cockroachdb.2018-08-01T20_24_51Z.001562.log.gz](https://github.com/cockroachdb/cockroach/files/2267283/cockroach.srv5-dc.cockroachdb.2018-08-01T20_24_51Z.001562.log.gz)
[cockroach.srv5-sv.cockroachdb.2018-08-01T20_24_38Z.001537.log.gz](https://github.com/cockroachdb/cockroach/files/2267284/cockroach.srv5-sv.cockroachdb.2018-08-01T20_24_38Z.001537.log.gz)
[cockroach.srv6-ch.cockroachdb.2018-08-01T20_37_07Z.001580.log.gz](https://github.com/cockroachdb/cockroach/files/2267285/cockroach.srv6-ch.cockroachdb.2018-08-01T20_37_07Z.001580.log.gz)
[cockroach.srv6-dc.cockroachdb.2018-08-01T20_38_27Z.001474.log.gz](https://github.com/cockroachdb/cockroach/files/2267286/cockroach.srv6-dc.cockroachdb.2018-08-01T20_38_27Z.001474.log.gz)
[cockroach.srv6-sv.cockroachdb.2018-08-01T20_35_44Z.001515.log.gz](https://github.com/cockroachdb/cockroach/files/2267287/cockroach.srv6-sv.cockroachdb.2018-08-01T20_35_44Z.001515.log.gz)
[cockroach.srv7-ch.cockroachdb.2018-08-01T20_53_59Z.001487.log.gz](https://github.com/cockroachdb/cockroach/files/2267288/cockroach.srv7-ch.cockroachdb.2018-08-01T20_53_59Z.001487.log.gz)
[cockroach.srv7-dc.cockroachdb.2018-08-01T20_55_17Z.019673.log.gz](https://github.com/cockroachdb/cockroach/files/2267289/cockroach.srv7-dc.cockroachdb.2018-08-01T20_55_17Z.019673.log.gz)
[cockroach.srv7-sv.cockroachdb.2018-08-01T20_53_50Z.001459.log.gz](https://github.com/cockroachdb/cockroach/files/2267290/cockroach.srv7-sv.cockroachdb.2018-08-01T20_53_50Z.001459.log.gz)
You were looking at [srv8-sv](https://github.com/cockroachdb/cockroach/files/2258256/cockroach.log) before
[cockroach.srv8-ch.cockroachdb.2018-08-01T21_35_00Z.001507.log.gz](https://github.com/cockroachdb/cockroach/files/2267291/cockroach.srv8-ch.cockroachdb.2018-08-01T21_35_00Z.001507.log.gz)
[cockroach.srv8-dc.cockroachdb.2018-08-01T21_35_13Z.001535.log.gz](https://github.com/cockroachdb/cockroach/files/2267292/cockroach.srv8-dc.cockroachdb.2018-08-01T21_35_13Z.001535.log.gz)

**Comment by tschottdorf** [comment]
Thanks, @jethrogb. I've taken a look at the logs yet am unable to identify the root cause from them. I opened PR #28519 which will help identify these issues better in v2.1, and filed issue #28517 (which is something probably unrelated I saw in one of your logs).

Are you running with replication factor 3 or 5?
If this problem reoccurs, could you share a screenshot of `/#/reports/range/1`?

**Comment by jethrogb** [comment]
Thanks for looking into this. I'm using a replication factor of 15. If it happens again, I'll try to report on the ranges.

**Comment by tschottdorf** [comment]
Just to double check, you're using a replication factor of 15? This would mean that you keep 15 copies of each bit of data, i.e. one copy on each of the nodes. Is that correct?

**Comment by jethrogb** [comment]
That's right. It seems a bit excessive to me too, but that's what it's at right now (according to `cockroach zone get`).

**Comment by petermattis** [comment]
I'm fairly confident that this is a duplicate of #27731. The TLDR is that single node restarts can cause a partition in the gossip network due the sentinel key being gossiped on both sides of the gossip partition which can occur in 2.0.x because the leaseholder for the range 1 lease can switch back and forth between nodes on either side of the gossip partition.
